ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -665,6 +665,69 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
     }
   }
 
+  @Override
+  public void forceDelete(List<MessageInfo> infosToDelete) throws StoreException {
+    checkStarted();
+    checkDuplicates(infosToDelete);
+    final Timer.Context context = metrics.deleteResponse.time();
+    try {
+      Offset indexEndOffsetBeforeCheck = index.getCurrentEndOffset();
+      // If the PubBlob exists, we are supposed to call BlobStore.delete.
+      // forceDelete is supposed to use when the blob doesn't exist.
+      for (MessageInfo info : infosToDelete) {
+        IndexValue value =
+            index.findKey(info.getStoreKey(), new FileSpan(index.getStartOffset(), indexEndOffsetBeforeCheck));
+        if (value != null) {
+          throw new StoreException("Cannot force delete id " + info.getStoreKey() + " since the id exists",
+              StoreErrorCodes.Already_Exist);
+        }
+      }
+      synchronized (storeWriteLock) {
+        List<InputStream> inputStreams = new ArrayList<>(infosToDelete.size());
+        List<MessageInfo> updatedInfos = new ArrayList<>(infosToDelete.size());
+
+        for (MessageInfo info : infosToDelete) {
+          MessageFormatInputStream stream =
+              new DeleteMessageFormatInputStream(info.getStoreKey(), info.getAccountId(), info.getContainerId(),
+                  info.getOperationTimeMs(), info.getLifeVersion());
+          updatedInfos.add(
+              new MessageInfo(info.getStoreKey(), stream.getSize(), info.getAccountId(), info.getContainerId(),
+                  info.getOperationTimeMs(), info.getLifeVersion()));
+          inputStreams.add(stream);
+        }
+
+        Offset endOffsetOfLastMessage = log.getEndOffset();
```

Comment:
we should call index.findKey again with the filespan being the old end offset to the new end offset just to make sure there is no new entry is added while we acquire the lock.

Reply:
Nice catch. I missed that when I read the corresponding delete function implementation. Also added unit test for it. Thanks,